### ISSUE TITLE
`pre-commit/mirrors-eslint` uses more than twice the memory on pre-commit.ci than locally resulting in OOM

### ISSUE THREAD

**Issue by SamuelT-Beslogic** [pre-commit-ci/issues#229]
Follow-up to: https://github.com/pre-commit-ci/issues/issues/223
Concerning the same PR: https://github.com/BesLogic/releaf-canopeum/pull/177

I've been trying to identify what makes the CI run use so much memory that it results in Out Of Memory errors, but I can't replicate locally.

I can run pre-commit locally with a 704MiB restriction just fine:
`$Env:NODE_OPTIONS="--max-old-space-size=704"; pre-commit run --all`

But on the CI, it busts somewhere above 1900MiB:
https://results.pre-commit.ci/run/github/771018559/1724365854.9aT1lXiZTW6V5XzoNwPZvA
```
<--- Last few GCs --->

[64:0x6e91290]    44330 ms: Mark-sweep (reduce) 1886.7 (1924.6) -> 1886.2 (1925.1) MB, 1281.6 / 0.0 ms  (average mu = 0.086, current mu = 0.029) allocation failure; scavenge might not succeed
[64:0x6e91290]    45639 ms: Mark-sweep (reduce) 1887.3 (1925.1) -> 1886.7 (1925.6) MB, 1303.6 / 0.0 ms  (average mu = 0.046, current mu = 0.004) allocation failure; scavenge might not succeed
```

I've made sure ESLint only runs where it needs to (see `args` in my config)

I've disabled my 10 slowest rules as reported by ESLint when setting `TIMING=1` without any noticeable change locally or on CI (so it seems the issue isn't directly related).

I'm a bit at a loss of what to test right now, I was hoping that despite the vastly different memory usage between local and CI if I could at least reduce the usage locally it would reflect on the CI.
My next guess is maybe something about the project layout in the CI causes some plugins to do extra analysis and store extra stuff in memory (like, let's say, typescript-eslint's parser creating more typescript programs). But making our shared configs able to run without `@typescript-eslint` just to test that hypothesis is quite the big change)

So I'm opening this issue to at least report what is, to me, unexpected memory usage for `pre-commit/mirrors-eslint` on `pre-commit.ci`

<details>
<summary><code>.pre-commit-config.yaml`</code></summary>

```yaml
# You can run this locally with `pre-commit run [--all]`
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.6.0
    hooks:
      - id: trailing-whitespace
        args: [--markdown-linebreak-ext=md]
      - id: end-of-file-fixer
      - id: mixed-line-ending
        args: [--fix=lf]
      - id: check-yaml
      - id: check-toml
      - id: check-merge-conflict
      - id: check-case-conflict

  # You can run this locally with `ruff check --fix || ruff format`
  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.5.7 # must match canopeum_backend/pyproject.toml
    hooks:
      # Run the linter.
      - id: ruff
        args: [--fix]
      # Run the formatter.
      - id: ruff-format

  # You can run this locally with `npm run lint:fix`
  - repo: https://github.com/pre-commit/mirrors-eslint
    rev: v8.56.0
    hooks:
      - id: eslint
        args:
          [
            canopeum_frontend,
            --ignore-path,
            canopeum_frontend/.gitignore,
            --fix,
            --quiet,
          ]
        files: \.[jt]sx?$ # *.js, *.jsx, *.ts and *.tsx
        types: [file]
        additional_dependencies: [
            # Keep these in sync with canopeum_frontend/package.json
            "eslint-config-beslogic@2.4.5",
            "eslint-plugin-jsx-a11y@6.8.0",
            "eslint-plugin-react@7.34.1",
            "eslint-plugin-react-hooks@4.6.0",
            "eslint-plugin-react-prefer-function-component@3.3.0",
            "eslint-plugin-react-refresh@0.4.5",
            # NOTE: Optional, removing to reduce bundle size: exceeds tier max size 250MiB
            # "eslint-plugin-total-functions",
            # "eslint-plugin-no-autofix",
            # "eslint-plugin-testing-library",
            # "eslint-plugin-jest-formatting",
            # "eslint-plugin-jest",
            # "eslint-plugin-etc",
            #
            # pre-commit required *explicitly* setting all dependencies, which is... annoying
            # (this is to prevent inconsistent dependency ranges and promote repeatability)
            # Dependencies of ESLint-Config-Beslogic
            "@eslint-community/eslint-plugin-eslint-comments@4.1",
            "confusing-browser-globals@1.0",
            "eslint@8.56.0",
            "eslint-import-resolver-typescript@3.2.6",
            "eslint-plugin-autofix@2.0",
            "eslint-plugin-extra-rules@0.0.0-development",
            "eslint-plugin-import@2.27.3",
            "eslint-plugin-no-relative-import-paths@1.5.3",
            "eslint-plugin-prefer-arrow@1.2",
            "eslint-plugin-regexp@2.0",
            "eslint-plugin-simple-import-sort@12.0",
            "eslint-plugin-sonarjs@1.0",
            "eslint-plugin-unicorn@49.0",
            "eslint-plugin-unused-imports@3.2",
            # Transitive dependencies of ESLint-Config-Beslogic
            "@typescript-eslint/eslint-plugin@7.18", # dependency of typescript-eslint
            "@typescript-eslint/parser@7.18", # dependency of typescript-eslint
            "semver@6.3.1", # dependency of eslint-plugin-import
          ]

ci:
  autoupdate_schedule: quarterly
```

</details>

**Comment by asottile** [comment]
how many vcpus do you have available?  there's 4 available which would explain the quadruple of memory (and oom)

**Comment by asottile** [comment]
the docker image is available in every run so you should be able to reproduce and debug there as well

**Comment by SamuelT-Beslogic** [comment]
> how many vcpus do you have available? there's 4 available which would explain the quadruple of memory (and oom)

If I understand the question correctly: 16 logical processors as Task manager calls them
![image](https://github.com/user-attachments/assets/6c5fee7a-814e-4788-b739-e2d255efad36)

> the docker image is available in every run so you should be able to reproduce and debug there as well

Running `nproc` in the docker image gives `16`

Extremely slow since running in the docker image from Windows is WSL2 based (and purposefully choking the memory allocation likely doesn't help), but the following still works:
```shell
> docker run -v .:/project_root -it ghcr.io/pre-commit-ci/runner-image:2024-05-11-4bb81b2 bash
root@702fe3b54123:/# cd project_root
root@702fe3b54123:/project_root# NODE_OPTIONS=--max-old-space-size=704 && pre-commit run --all
```

---

Oh and just to make sure it wasn't some oddity with the mirror-eslint hook, I tried as a local repo hook with the same result
```yaml
- repo: local
    hooks:
      - id: eslint
        name: eslint
        language: node
        entry: bash -c 'cd canopeum_frontend && npm run eslint -- --fix'
        files: \.[jt]sx?$ # *.js, *.jsx, *.ts and *.tsx
        types: [file]
        additional_dependencies: [...IDEM]
```

**Comment by asottile** [comment]
comparing ram on windows against ram on linux isn't really going to be meaningful

also this sets the environment variable and then doesn't use it in the command: 

```
NODE_OPTIONS=--max-old-space-size=704 && pre-commit run --all
```

you probably want

```
NODE_OPTIONS=--max-old-space-size=704 pre-commit run --all
```

**Comment by SamuelT-Beslogic** [comment]
I managed to work my way through it, here's some useful information if someone else has the same issue:

With only my JavaScript preset, the rule [`import/no-unresolved`](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-unresolved.md) may have been the sole culprit.

With my JS+TS presets, that rule was already disabled (since TypeScript already validates imports).

Adding my React presets, I had to use the experimental project service:
```js
/** @type {import("eslint").Linter.Config} */
module.exports = {
  parserOptions: {
    // project: ['tsconfig?(.*).json'],
    // project: true,
    // Experimental, but causes pre-commit to actually find the project and avoids OOM
    EXPERIMENTAL_useProjectService: true,
  },
}
```
(related: https://github.com/typescript-eslint/typescript-eslint/issues/1192#issuecomment-2105896680)

---

I've has a couple plugins act weirdly under `pre-commit.ci`, namely https://github.com/aladdin-add/eslint-plugin/issues/98; and https://www.npmjs.com/package/eslint-import-resolver-typescript not resolving shared tsconfigs.json in node_modules. So I guess my two OOM causes above were more symptoms of those path/module resolution issues with ESLint under pre-commit.ci.
